crates/wasi-common/src/wasi.rs

```diff
@@ -2,118 +2,474 @@
+    ) -> std::result::Result<(), types::Errno> {
+        unimplemented!("path_remove_directory")
+    }
+
+    fn path_rename(
+        &mut self,
+        _fd: types::Fd,
+        _old_path: &GuestString<'_>,
+        _new_fd: types::Fd,
+        _new_path: &GuestString<'_>,
+    ) -> std::result::Result<(), types::Errno> {
+        unimplemented!("path_rename")
+    }
+
+    fn path_symlink(
+        &mut self,
+        _old_path: &GuestString<'_>,
+        _fd: types::Fd,
+        _new_path: &GuestString<'_>,
+    ) -> std::result::Result<(), types::Errno> {
+        unimplemented!("path_symlink")
+    }
+
+    fn path_unlink_file(
+        &mut self,
+        _fd: types::Fd,
+        _path: &GuestString<'_>,
+    ) -> std::result::Result<(), types::Errno> {
+        unimplemented!("path_unlink_file")
+    }
+
+    fn poll_oneoff(
+        &mut self,
+        _in_: GuestPtr<types::Subscription>,
+        _out: GuestPtrMut<types::Event>,
+        _nsubscriptions: types::Size,
+    ) -> std::result::Result<types::Size, types::Errno> {
+        unimplemented!("poll_oneoff")
+    }
+
+    fn proc_exit(&mut self, _rval: types::Exitcode) -> std::result::Result<(), ()> {
+        unimplemented!("proc_exit")
+    }
+
+    fn proc_raise(&mut self, _sig: types::Signal) -> std::result::Result<(), types::Errno> {
+        unimplemented!("proc_raise")
+    }
+
+    fn sched_yield(&mut self) -> std::result::Result<(), types::Errno> {
+        unimplemented!("sched_yield")
+    }
+
+    fn random_get(
+        &mut self,
+        _buf: GuestPtrMut<u8>,
+        _buf_len: types::Size,
+    ) -> std::result::Result<(), types::Errno> {
+        unimplemented!("random_get")
+    }
+
+    fn sock_recv(
+        &mut self,
+        _fd: types::Fd,
+        _ri_data: &types::IovecArray<'_>,
+        _ri_flags: types::Riflags,
+    ) -> std::result::Result<(types::Size, types::Roflags), types::Errno> {
+        unimplemented!("sock_recv")
+    }
+
+    fn sock_send(
+        &mut self,
+        _fd: types::Fd,
+        _si_data: &types::CiovecArray<'_>,
+        _si_flags: types::Siflags,
+    ) -> std::result::Result<types::Size, types::Errno> {
+        unimplemented!("sock_send")
+    }
+
+    fn sock_shutdown(
+        &mut self,
+        _fd: types::Fd,
+        _how: types::Sdflags,
+    ) -> std::result::Result<(), types::Errno> {
+        unimplemented!("sock_shutdown")
+    }
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, thiserror::Error)]
+#[error("{}", self.0)]
+pub struct Error(types::Errno);
+
+impl From<types::Errno> for Error {
+    fn from(errno: types::Errno) -> Self {
+        Self(errno)
+    }
+}
+
+impl From<std::io::Error> for Error {
+    fn from(err: std::io::Error) -> Self {
+        log::error!("I/O error: {:?}", err);
+        Self::from_raw_os_error(err.raw_os_error().unwrap())
+    }
+}
+
+pub(crate) trait FromRawOsError {
+    fn from_raw_os_error(code: i32) -> Self;
+}
+
```

Comment:
My feedback on this Error type: I would prefer an `enum Error { Io(std::io::Error), Errno(types::Errno) }` and you can use `thiserror` to implement the printing and `#[from]` on both of those inner field to get the From impl

Reply:
I'm actually looking at some changes to help make it easier to support multiple versions of WASI snapshots, and having WASI errno values exposed in the API that normal embedding code uses makes this difficult. This has led me to think that we want something different from both of these.

My idea is that embedding APIs like `WasiCtxBuilder` should return plain `io::Error` errors, or possibly `io::Error` and some `WasiCtxBuilder`-specific error type, because it's called from host code which doesn't have any reason to know about WASI errno values, and where being dependent on a specific WASI API version is inconvenient. And then, of course, calling into the WASI implementation functions should just return a WASI errno, and not an `io::Error`.

Does that make sense? If so, I can submit some PRs that refactor the code in that direction.

Reply:
Hmm, I actually wanted to suggest getting rid of the problem of storing intermediate error types and simply log if such an error occurred and immediately convert to the correct errno value, but I'm open to suggestions here. Originally we've had a couple of levels of indirection and it proves messy and difficult to manage, so hence my suggestion of removing those.

@sunfishcode Hmm, is `io::Error` what we want though? I thought that it was OK for us to deal with raw WASI errno values at this level in the implementation. However, I can see the benefits of letting Rust take care of that for us, especially since then we don't have to deal with error conversions between different hosts and WASI (looking at you here, Windows). @pchickey I reckon returning `io::Error` should be fine in `wiggle`. We would have trait fns return `Result<result_type, io::Error>`, and underneath, `wiggle` would automatically deal with the conversion from `io::Error` to the matching `types::Errno` value.

@sunfishcode while we're here, a word of caution if we want to go the `io::Error` route. We should somehow factor in the cases where we'd want to return `std::ffi::NulError`, `std::str::Utf8Error` and `std::num::TryFromIntError` which can also happen in `wasi-common`.

Reply:
Now that I think about it some more, I guess we wouldn't want to introduce any conversion logic between host and guest into `wiggle` since I guess `wiggle` should be host-independent (i.e., only worry about the guest). I guess we could add this layer in-between `wiggle` and `wasi-common` hostcalls which would then be invoked in the trait impls in `wiggle`. Does that make sense?

Reply:
I'm all for keeping the wasi types::Errno enum exposed only at the WASI interfaces themselves, and having wasi-common use some other, more stable error enum to expose to other parts of the host program.

The UTF8Error and TryFromIntError and friends are part of the implementation details of the wasi-common host calls, so those should get converted to a types::Errno before exposing those to the guest. I don't know exactly how we should expose those to the host, you can see in the wiggle tests I mocked a design that just keeps a log (& eprintln!'s) of the GuestErrors https://github.com/kubkon/wiggle/blob/master/crates/test/src/lib.rs#L97

I think I will want some way to access those GuestErrors (https://github.com/kubkon/wiggle/blob/master/crates/runtime/src/error.rs) because they will be valuable for debugging guest programs. And, they should be a lot more stable than wasi's Errno. Ultimately they're designed for expressing different things to different users. GuestError is super descriptive and should be used for the guest code author to understand bugs in the system, but Errno is the value we actually pass to the guest program to indicate an invalid operation happened.

Reply:
@sunfishcode I've had a quick look at hiding WASI errno from `wasi-common` code, and I cannot see how we could *only* use `io::Error` since in several places we return very specific errno values which we cannot represent with *platform-independent* `io::Error`. I was thinking of encoding everything as raw OS error values but then they are not uniform across different platforms (*nix vs Windows for instance).

This is not to say I don't like the idea of using an error type that is independent from WASI errno and only converted to an appropriate error code when returning to the guest, but I reckon we need a little bit more than just reusing `io::Error`. Perhaps a custom `enum Error` which would be a superposition of `io::Error`, `num::TryFrom...`, **and** empty tags such as `BadFileDescriptor`, etc.? We'd end up with something like this:

```rust
#[derive(thiserror::Error)]
enum Error {
    #[error("bad file descriptor")]
    BadFileDescriptor,
    #[error("I/O error: {0}")]
    Io(#[from] std::io::Error),
    // ...
}
```

Then, we'd have the conversion to WASI errno auto-generated in `wig`/`wiggle`.

Is that something you've had in mind? Btw, I'm happy to help with this refactoring effort since it seems like quite a bit of work.

Reply:
OK, so to keep everyone in the loop here, I've had an offline chat with @sunfishcode, and we agreed that it's best to hide the fact that WASI's errno exists from the WASI syscall implementation, and then have the conversion from some higher-level error type (which you could think of as an error struct on the same level as `io::Error`) to the errno value auto-generated by `wig`/`wiggle`. I reckon this is an excellent suggestion by @sunfishcode which should future-proof `wasi-common` from future changes to the WASI ABI. Now, the plan is for @sunfishcode and myself to suggest such an error type and refactor `wasi-common` to use it. This should greatly simplify what is going on in this PR and generally ease our transition from `wig` to `wiggle`.